/*
This file holds general geometric types, for more complex messages.
*/

edition = "2023";

// TODO: Standardize this better? Should There be IntVal and DoubleVal,
// with value and units?

/* 2D unsigned integer point. */
message Point2u {
  uint32 x = 1;
  uint32 y = 2;
}

/* 2D double point */
message Point2d {
  double x = 1;
  double y = 2;
}

/* 2D unsigned integer size. */
message Size2u {
  uint32 x = 1;
  uint32 y = 2;
}

/* 2D double size. */
message Size2d {
  double x = 1;
  double y = 2;
}

/* 3D unsigned integer size. */
message Size3u {
  uint32 x = 1;
  uint32 y = 2;
  uint32 z = 3;
}

/* ROI Rectangle. */
message Rect2d {
  Point2d top_left = 1;
  Size2d size = 2;
}

/* ROI Rectangle, integer. */
message Rect2i {
  Point2u top_left = 1;
  Size2u size = 2;
}

/* Rotated Rectangle. */
message RotRect2d {
  Point2d top_left = 1;
  Size2d size = 2;
  sint32 angle = 3; // Rotation angle in degrees.
}

/* Rotated Rectangle, integer. */
message RotRect2i {
  Point2u top_left = 1;
  Size2u size = 2;
  sint32 angle = 3; // Rotation angle in degrees.
}
